### PR TITLE
[TRANSFORMATIONS] Fix Optional to match even with no inputs

### PR DESCRIPTION
**Pull request by CuriousPanCake**
[TRANSFORMATIONS] Fix Optional to match even with no inputs

### Details:
The Optional pattern type may create a wrong pattern to match if no inputs are provided to the Optional node. If no inputs present to the Optional type, it will not create an alternative branch(es) to check against resulting in the incorrect matching.

Fix that by adding a check for the number of inputs being 0.

Do a minor refactoring/renaming for the readability purposes.

### Tickets:
 CSV-133523

Signed-off-by: Andrii Staikov <andrii.staikov@intel.com>

### CODE REVIEW

File: src/core/src/pattern/op/optional.cpp [hunk 1]
```diff
@@ -15,11 +15,17 @@ std::vector<ov::DiscreteTypeInfo> ov::pass::pattern::op::Optional::get_optional_
...
-    or_in_values.push_back(wrap_node);
+    // Turn the Optional node into WrapType node to create a case where the Optional node is present
+    ov::OutputVector input_values_to_optional = input_values();
+    size_t num_input_values_to_optional = input_values_to_optional.size();
+    auto wrap_node =
+        std::make_shared<ov::pass::pattern::op::WrapType>(optional_types, m_predicate, input_values_to_optional);
 
-    if (matcher->match_value(std::make_shared<ov::pass::pattern::op::Or>(or_in_values), graph_value)) {
+    // Add the newly created WrapType node to the list containing its inputs and create an Or node with the list
+    input_values_to_optional.push_back(wrap_node);
```
Comment (itikhono) [inline src/core/src/pattern/op/optional.cpp:25]: maybe we can make it clearer using something like that?
```
 // we should use only "data" (0th) input of pattern::Optional in Or
auto pattern = num_input_values_to_optional == 0? wrap_node :  std::make_shared<ov::pass::pattern::op::Or>({wrap_node, input_values_to_optional[0]});

if (matcher->match_value(pattern, ...
```

Comment (itikhono) [reply src/core/src/pattern/op/optional.cpp:25]: and using std; and using ov::pass::pattern::op at the begging of file to simplify 

std::make_shared<ov::pass::pattern::op::Or> -> make_shared<Or>

Comment (CuriousPanCake) [reply src/core/src/pattern/op/optional.cpp]: Fixed

File: src/core/tests/pattern.cpp [hunk 2]
```diff
@@ -508,6 +510,68 @@ TEST(pattern, matching_optional) {
...
+    auto model_input2 = std::make_shared<op::v0::Parameter>(element::i32, shape);
+    auto model_add = std::make_shared<op::v1::Add>(model_input1->output(0), model_input2->output(0));
+    auto model_relu = std::make_shared<op::v0::Relu>(model_add->output(0));
+
+    auto pattern_add = ov::pass::pattern::optional<op::v1::Add>();
+    auto pattern_relu = std::make_shared<op::v0::Relu>(pattern_add->output(0));
+
+    TestMatcher tm;
+
+    ASSERT_TRUE(tm.match(pattern_relu, model_relu));
```
Comment (itikhono) [inline src/core/tests/pattern.cpp:525]: I would recommend to test pattern_map in these tests
Please check that pattern_map contains pattern_add/pattern_relu keys and the corresponding values are equal to model_add and model_relu.

We can do it in the next PR, not a problem.

File: src/core/tests/pattern.cpp [hunk 3]
```diff
@@ -508,6 +510,68 @@ TEST(pattern, matching_optional) {
...
+    auto model_input2 = std::make_shared<op::v0::Parameter>(element::i32, shape);
+    auto model_add = std::make_shared<op::v1::Add>(model_input1->output(0), model_input2->output(0));
+    auto model_relu = std::make_shared<op::v0::Relu>(model_add->output(0));
+
+    auto pattern_relu = ov::pass::pattern::optional<op::v0::Relu>();
+    auto pattern_relu1 = std::make_shared<op::v0::Relu>(pattern_relu->output(0));
+
+    TestMatcher tm;
+
+    ASSERT_TRUE(tm.match(pattern_relu1, model_relu));
```
Comment (itikhono) [inline src/core/tests/pattern.cpp:540]: please check that pattern_map[pattern_relu] is nullptr in this case

File: src/core/tests/pattern.cpp [hunk 4]
```diff
@@ -508,6 +510,68 @@ TEST(pattern, matching_optional) {
...
+}
+
+TEST(pattern, optional_testing) {
```
Comment (itikhono) [inline src/core/tests/pattern.cpp:543]: in OV we have some operations with "optional" inputs, e.g. Interpolate-11 operation
https://docs.openvino.ai/2023.3/openvino_docs_ops_image_Interpolate_11.html#

image

optional class might be used to match Interpolate with different number of inputs: 2 inputs and 3 inputs using 1 instance:
Interpolate(wrap_type, wrap_type, optional)

could you check if we have such a test? if not could you add it?

File: src/core/src/pattern/op/optional.cpp [hunk 5]
```diff
@@ -8,18 +8,52 @@
...
+   ┌──────────────┐                └──────┬───────┘
+   │    Relu      │                       │
+   └──────┬───────┘                ┌──────┴───────┐        ┌──────────────┐
+          │                        │WrapType<Relu>│        │     Relu     │
+   ┌──────┴───────┐                └──────┬───────┘        └───────┬──────┘
+   │Optional<Relu>│  Unfolds into         │                        │
+   └──────┬───────┘                       └────────┐      ┌────────┘
+          │                                        │      │
+        ┌─┴─┐                                     ┌┴──────┴┐
+        │ABS│                                     │   Or   │
```
Comment (itikhono) [inline src/core/src/pattern/op/optional.cpp:25]: minor: ABS -> Abs

**Pull request merged**